### ISSUE TITLE
Consumer parallelism?

### ISSUE THREAD

**Issue by eliaslevy** [joekiller/jruby-kafka#32]
Maybe I am being dense, but client API seems rather odd to me.  `Kafka::Group#run` takes a number of threads.  At first glance this may give you the idea that it will allow for parallelism when processing the messages.  But `Kafka::Group#run` also takes a queue, and when it creates the thread pool is creates a `Kafka::Consumer` for each thread and calls `Kafka::Consumer#run` on each.  `Kafka::Consumer#run` in turn just pushed messages it receives into the queue, rather than say, taking a `Proc` to process the messages with.

So what is the point of `Kafka::Group#run` having a number of threads argument if all messages will just be pushed into a single queue?  Passing a value other than 1 will make no difference.  If the user wants some parallelism they will have to create their own thread pool to consume from the queue.  They will also have to recreate the partitions provided by the Kafka streams.

Also, since the call to `Kafka::Group#run` just queues the messages, they will be considered processed by the Java consumer client, and thus potentially have their offsets committed, whether or not the messages have been truly handled by the code consuming messages off the queue, thus potentially resulting in the loss of all messages in the queue if the client dies.

**Comment by joekiller** [comment]
I think you make good points. Sending in a proc for each thread could work well.

Considering the current design, using more fetcher threads with a single thread safe message queue enables more flexibility in ensuring that the processor of messages always had a steady stream as a fetcher can only work on one partition at a time. If one partition is going slow, other threads are still able to fill the sink.

You last point is valid as well. These changes could work well.

**Comment by joekiller** [comment]
A little more on the third point. If you had a series of processes which were handling the messages on each fetcher thread, you could still lose a message in the queue if the processing function failed in some fashion after dequeing the message because from the Kafka consumer that message was read. Consumer offset auto commit defaults to enabled and at 60 second intervals. It's much more likely that you would crash and not have committed yet versus crashing and losing internal queued messages. Running very low offset intervals increases the load on your zookeeper nodes which may bottleneck the Kafka client.

Regardless, I like the idea of getting the processing closer to the fetcher threads. Some of this API reflects the nature of it's original primary use: passing messages off Kafka to a logstash managed queue.

**Comment by eliaslevy** [comment]
Re: more threads.  Apologies if I am wrong, as am just digging into Kafka, but in my reading of the Kafka Java/Scala consumer it appears that the code internally creates one thread per partition to fetch messages (a [`ConsumerFetcherThread`](https://github.com/apache/kafka/blob/43b92f8b1ce8140c432edf11b0c842f5fbe04120/core/src/main/scala/kafka/consumer/ConsumerFetcherThread.scala)), so there is no blocking the processor on a slow partition.

The `ConsumerFetcherThread`s in turn write the fetched messages into a queue stored in the [`PartitionTopicInfo`](https://github.com/apache/kafka/blob/43b92f8b1ce8140c432edf11b0c842f5fbe04120/core/src/main/scala/kafka/consumer/PartitionTopicInfo.scala) objects.  These queues are also stored within the `KafkaStream` objects.  And it is from them that messages fetched via the iterator.

It would seem that if you call `createMessageStreamsByFilter` with a map for a topic with a value of 1, what you end up with is one thread per partition writing to a single queue stored in the single returned `KafkaStream` object. 

Basically, the Scala/Java client appears to be doing the same work you are recreating in Ruby of multiplexing messages from multiple partitions into a single queue.  

Passing a value higher than 1 to `Kafka::Group#run` adds no parallelism.  The Java Kafka client will create the same number of fetcher threads (one per partition).  It will merely result in having multiple `KafkaStream`s, each with its own queue, to which different partition are multiplexed into, and then having the Ruby code multiplex those further into a single queue.

**Comment by eliaslevy** [comment]
Re: loosing messages. You are right.  You'd loose messages either way with auto commit turned on.  At first I thought the Kafka client was keeping track of processed messages, but as you point out it just commit them on an interval when auto commit if on, which is pretty bad as there is no guarantee the message will indeed have been processed by the time the offset if committed.

Obviously the solution is to turn auto commit off and perform commits manually.  For this you would have to expose [`commitOffsets`](https://github.com/apache/kafka/blob/3df46bf4ce9c134cc7b532be3d01f920127be706/core/src/main/scala/kafka/javaapi/consumer/ConsumerConnector.java#L76) in Ruby.

The Ruby code could also take care of this.  For instead, if the consumer took a Proc to process messages, it could assume that any message for which the Proc did not raise an exception was processed successfully.  It could then track the highest offset per partition processed successfully and call `commitOffsets` at some interval to commit processed offsets.

**Comment by joekiller** [comment]
As I said before, I see the merit in what you are stating. I'm open to proposed changes. I think this issues was masked to me prior because I almost always scale out my processes with single threads on small EC2 instances so they scale with load, which basically masked this issue.

**Comment by eliaslevy** [comment]
See #33 for a proposal.

**Comment by eliaslevy** [comment]
Closing as #33 was merged.
